pkg/certificate/mrc_reconciler.go

```diff
@@ -1,214 +1,228 @@
+			return false, nil
 		}
+		// update since the issuers do not match the desired MRCs
+		return true, nil
 	}
 
-	return m.mrcClient.UpdateMeshRootCertificate(mrc)
+	return true, nil
 }
 
-// isTerminal checks if the MRC is in a terminal state.
-func isTerminal(mrc *v1alpha2.MeshRootCertificate) bool {
-	// TODO(5046): check if the mrc is in a terminal state.
-	return true
+func (m *Manager) updateIssuers(signingIssuer, validatingIssuer *issuer) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.signingIssuer = signingIssuer
+	m.validatingIssuer = validatingIssuer
+	log.Trace().Msgf("setting issuers for validating[%s] and signing[%s]", validatingIssuer.ID, signingIssuer.ID)
+	return nil
 }
 
-// state and condition get updated together.
-func (m *Manager) updateMRCState(mrc *v1alpha2.MeshRootCertificate) error {
-	// TODO(5046): update the MRC state & conditions, and issue the update via the MRCClient.
+func (m *Manager) getCertIssuers(desiredSigningMRC, desiredValidatingMRC *v1alpha2.MeshRootCertificate) (*issuer, *issuer, error) {
+	if desiredSigningMRC == nil || desiredValidatingMRC == nil {
+		log.Error().Err(ErrUnexpectedNilMRC).Msg("unexpected nil MRC provided when getting cert issuers from MRCs")
+		return nil, nil, ErrUnexpectedNilMRC
+	}
 
-	// If it's not in a terminal state, set the next transition time.
-	// NOTE: consider that this isTerminal check will need to apply against the updated mrc state which should happen
-	// above in this method.
-	if isTerminal(mrc) {
-		mrc.Status.TransitionAfter = nil
+	var desiredSigningIssuer, desiredValidatingIssuer *issuer
+	desiredSigningIssuer, err := m.getCertIssuer(desiredSigningMRC)
+	if err != nil {
+		return nil, nil, err
+	}
+	// don't get the issuer again if there is a single MRC in the control plane
+	if desiredSigningMRC == desiredValidatingMRC {
+		desiredValidatingIssuer = desiredSigningIssuer
 	} else {
-		mrc.Status.TransitionAfter = &metav1.Time{Time: time.Now().Add(MrcDurationPerStage)}
+		desiredValidatingIssuer, err = m.getCertIssuer(desiredValidatingMRC)
+		if err != nil {
+			return nil, nil, err
+		}
 	}
 
-	// TODO(5046): add the retry loop.
-	return m.mrcClient.UpdateMeshRootCertificate(mrc)
+	return desiredSigningIssuer, desiredValidatingIssuer, nil
 }
 
-func (m *Manager) shouldSetIssuers(mrc *v1alpha2.MeshRootCertificate) bool {
-	// TODO(5046): check the states, and if in some form of an active state, AND if the MRC is not already the existing
-	// then we should set MRC, so return true
-	return true
-}
+func (m *Manager) getCertIssuer(mrc *v1alpha2.MeshRootCertificate) (*issuer, error) {
+	m.mu.Lock()
+	signingIssuer := m.signingIssuer
+	validatingIssuer := m.validatingIssuer
+	m.mu.Unlock()
+
+	// if the issuer has already been created for the specified MRC,
+	// return the existing issuer
+	if signingIssuer != nil && mrc.Name == signingIssuer.ID {
```

Comment:
if we have this check here, then we can be less defensive about the shouldUpdateIssuer, and remove the checks on whether the desired is already set or not

Reply:
I think `shouldUpdateIssuer` should keep the check if the issuers are already set to their desired values. If this condition is true, the issuers should not be updated and therefore `shouldUpdateIssuer` should return false.

Reply:
I am not sure if this handles the scenario `desiredSigningMRC != desiredValidatingMRC and both MRCs have active intents, their state is non-deterministic.` ?

Even if it does, I kind of like being explicit about though and exiting the reconciliation early.  It isn't immediately obvious that there are no changes otherwise.